### DISCUSSION TITLE
Blessing: Proposal to add new feature to repository.find using multiple where OR/AND clauses

### DISCUSSION THREAD

**Discussion by dash-tobin** [typedorm/typedorm#330] [Ideas]
From what I can tell, on repository.find->filter->where the OR/AND clause is limited to one value per field. Please tell me if I'm wrong. What I'm suggesting is create something like:

```
aws dynamodb query \
    --table-name Thread \
    --key-condition-expression "(#KY_CE_PK = :KY_CE_PK) and (begins_with(#KY_CE_SK, :KY_CE_SK))" \
    --expression-attribute-values  '{":KY_CE_PK": {"S": "ENTITY#1"}, ":KY_CE_SK":{"S":"account#"}, ":dA1": { "S": "A1"}, ":dB1": { "S": "B1"}, ":dA2": { "S": "A2"}, ":dB2": { "S": "B2"}}' \
    --filter-expression "(#FE_detailsA_1 = :dA1 and #FE_detailsB_1 = :dB1) or (#FE_detailsA_2 = :dA2 and #FE_detailsB_2 = :dB2)" \
    --expression-attribute-names '{"#KY_CE_PK":"PK","#KY_CE_SK":"SK","#FE_detailsA_1": "detailsA", "#FE_detailsB_1": "detailsB",,"#FE_detailsA_2": "detailsA", "#FE_detailsB_2": "detailsB"}' \
```

So if my table looks like:

| pk                   | sk                    | detailA           | detailB |
| ------------- |:-------------:| :-------------:|-------:|
| ENTITY#1      | account#A1#B1      | A1                 |  B1     |
| ENTITY#1      | account#A2#B2     | A2                | B2     |
| ENTITY#1      | account#A1#B2  |    A1             | B2     |
| ENTITY#1      | account#A2#B1    |    A2            | B1     |

Suggested query looks like this:
```
return getEntityManager().find(Pref, { id }, 
  { 
    where: 
      { 
        OR: [{ 
          AND: { 
            detailA: { EQ: 'A1' },
            detailB: { EQ: 'B1' },
        }, { 
          AND: {
            detailA: { EQ: 'A2' },
            detailB: { EQ: 'B2' },
          }
       }]
    },
    keyCondition: { 
      BEGINS_WITH: '${type}#' 
    }
 }
```

Result
| pk                   | sk                    | detailA           | detailB |
| ------------- |:-------------:| :-------------:|-------:|
| ENTITY#1      | account#A1      | A1                 |  B1     |
| ENTITY#2      | account#A2     | A2                | B2     |

This would also allow for multiple OR values on the same field like this:
```
return getEntityManager().find(Pref, { id }, 
  { 
    where: 
      { 
        OR: [{ 
          AND: { 
            detailA: { EQ: 'A1' },
        }, { 
          AND: {
            detailA: { EQ: 'A2' },
          }
       }]
    },
    keyCondition: { 
      BEGINS_WITH: '${type}#' 
    }
 }
```

This would also work in current implementation, so this would still work:
```
return getEntityManager().find(Pref, { id }, 
  { 
    where: 
      { 
        OR: {
            detailA: { EQ: 'A1' },
            detailB: { EQ: 'B1' },
        }, 
    },
    keyCondition: { 
      BEGINS_WITH: '${type}#' 
    }
 }
```
This was [the only documentation](https://github.com/typedorm/typedorm/blob/develop/docs/how-to-recipes.md#apply-filter-to-query) I could find.

**Comment by whimzyLive** [comment]
@dash-tobin 
It is already under discussion here https://github.com/typedorm/typedorm/issues/230
